README.md
Hi and welcome to team Gilded Rose.

As you know, we are a small inn with a prime location in a prominent city ran
by a friendly innkeeper named Allison.  We also buy and sell only the finest
goods. Unfortunately, our goods are constantly degrading in quality as they
approach their sell by date.

We have a system in place that updates our inventory for us. It was developed
by a no-nonsense type named Leeroy, who has moved on to new adventures. Your
task is to add the new feature to our system so that we can begin selling a
new category of items.

First an introduction to our system:

  - All items have a sell-in value which denotes the number of days we have to
    sell the item

  - All items have a quality value which denotes how valuable the item is

  - At the end of each day our system lowers both values for every item

Pretty simple, right? Well this is where it gets interesting:

  - Once the sell by date has passed, quality degrades twice as fast

  - The quality of an item is never negative

  - "Aged Brie" actually increases in quality the older it gets

  - The quality of an item is never more than 50

  - "Sulfuras", being a legendary item, never has to be sold or decreases in
    quality

  - "Backstage passes", like aged brie, increases in quality as it's sell-in
    value approaches; quality increases by 2 when there are 10 days or less
    and by 3 when there are 5 days or less but quality drops to 0 after the
    concert

We have recently signed a supplier of conjured items. This requires an update
to our system:

  - "Conjured" items degrade in quality twice as fast as normal items

Feel free to make any changes to the update-quality method and add any new code
as long as everything still works correctly. However, do not alter the item
function as that belongs to the goblin in the corner who will insta-rage and
one-shot you as he doesn't believe in shared code ownership.


Just for clarification, an item can never have its quality increase above 50,
however "Sulfuras" is a legendary item and as such its quality is 80 and it
never alters.


Additionally, this project was scaffolded with Leinengen. If you have Lein installed, you can run tests via `lein test`.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-05-30
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-05-30
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/gilded-rose/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gilded-rose/compare/0.1.0...0.1.1


test/gilded_rose/core_test.clj
;the specs

(ns gilded-rose.core-test
  (:require [clojure.test :refer [deftest testing is]]
            [gilded-rose.core :as core]))

(defn do-x-times [func input x]
  (nth (iterate func input) x))

; Running out of time, these could be cleaner and broken into individual tests
; or we could use Midje or Speclj
(deftest single-run
  (let [inventory (core/update-current-inventory)
        [vest brie elixir sulfuras tickets] inventory]
    (testing "sulfuras should maintain 80"
      (is (= 80 (:quality sulfuras))))
    (testing "brie should go up to 1"
      (is (= 1 (:quality brie))))
    (testing "tickets should go up to 21"
      (is (= 21 (:quality tickets))))
    (testing "elixir should go down to 6"
      (is (= 6 (:quality elixir))))
    (testing "vest should go down to 19"
      (is (= 19 (:quality vest))))))


(deftest multi-run
  (let [inventory (core/update-current-inventory)
        inventory (do-x-times core/update-inventory inventory 3)
        [vest brie elixir sulfuras tickets] inventory]
    (testing "sulfuras should maintain 80"
      (is (= 80 (:quality sulfuras))))
    (testing "brie should go up to 5"
      (is (= 5 (:quality brie))))
    (testing "tickets should go up to 24"
      (is (= 24 (:quality tickets))))
    (testing "elixir should go down to 3"
      (is (= 3 (:quality elixir))))
    (testing "vest should go down to 16"
      (is (= 16 (:quality vest))))))


; Probing at edge cases, these should be cleaner, refactor in the future
(deftest x-run
  (let [inventory (core/update-current-inventory)
        inventory (do-x-times core/update-inventory inventory 10)
        [vest brie elixir sulfuras tickets] inventory]
    (testing "sulfuras should maintain 80"
      (is (= 80 (:quality sulfuras))))
    (testing "brie should go up to 5"
      (is (= 19 (:quality brie))))
    (testing "tickets should go up to 36"
      (is (= 36 (:quality tickets))))
    (testing "elixir should not go below 0"
      (is (= 0 (:quality elixir))))
    (testing "vest should go down to 9"
      (is (= 9 (:quality vest))))))


(deftest un-brie-leviable
  (let [inventory (core/update-current-inventory)
        inventory (do-x-times core/update-inventory inventory 50)
        [vest brie elixir sulfuras tickets] inventory]
    (testing "brie should should go no higher than 50"
      (is (= 50 (:quality brie))))))


(deftest conjured
  (let [inventory
        [(core/item "+5 Dexterity Vest" 10 20)
         (core/item "Aged Brie" 2 0)
         (core/item "Elixir of the Mongoose" 5 7)
         (core/item "Sulfuras, Hand Of Ragnaros" 0 80)
         (core/item "Backstage passes to a TAFKAL80ETC concert" 15 20)]
        inventory (map #(assoc % :tags [:conjured]) inventory)
        inventory (core/update-inventory inventory)
        [vest brie elixir sulfuras tickets] inventory]
    (testing "sulfuras should maintain 80"
      (is (= 80 (:quality sulfuras))))
    (testing "brie should go up to 2"
      (is (= 2 (:quality brie))))
    (testing "tickets should go up to 22"
      (is (= 22 (:quality tickets))))
    (testing "elixir should go down to 5"
      (is (= 5 (:quality elixir))))
    (testing "vest should go down to 18"
      (is (= 18 (:quality vest))))))


src/gilded_rose/core.clj
(ns gilded-rose.core)

; Legendary Items don't change in quality
(def legendary-item-names
  '("Sulfuras, Hand Of Ragnaros"))

; I won't touch this function no matter how badly I want to.
(defn item [item-name, sell-in, quality]
  {:name item-name, :sell-in sell-in, :quality quality})


(defn- expired? [item]
  (< (:sell-in item) 0))


(defn- update-sell-in [item]
  (update item :sell-in dec))


; We can leverage update-quality for each item
; and just specify how certain ones (aged brie)
; deviate from the from the default.
(defmulti update-quality
  (fn [item]
    (:name item)))


(defmethod update-quality "Aged Brie"
  [item]
  (let [quality (:quality item)]
      (if (< quality 50)
        (if (expired? item)
          (update item :quality + 2)
          (update item :quality inc))
        (assoc item :quality 50))))


(defmethod update-quality "Backstage passes to a TAFKAL80ETC concert"
  [item]
  (let [quality (:quality item)]
      (if (< quality 50)
        (cond
          (expired? item)
            (assoc item :quality 0)
          (< (:sell-in item) 5)
            (update item :quality + 3)
          (< (:sell-in item) 10)
            (update item :quality + 2)
          :else
            (update item :quality inc))
        (assoc item :quality 50))))


(defmethod update-quality :default
  [item]
  (if-not (contains? (set legendary-item-names) (:name item))
    (cond
      (expired? item)
        (update item :quality - 2)
      :else
        (update item :quality dec))
    item))


(defn- handle-negative-qualities
  "Resets a quality to 0 provided it is below legal bounds"
  [item]
  (if (< (:quality item) 0)
    (assoc item :quality 0)
    item))


(defn- handle-conjured
  "If the item is conjured, we go ahead and trigger update quality once"
  [item]
  (if (some #{:conjured} (:tags item))
    (update-quality item)
    item))


(defn- update-item
  "pipeline for handline item updates w/ quality and sell-in"
  [item]
  (-> item
    handle-conjured
    update-quality
    update-sell-in
    handle-negative-qualities))


(defn update-inventory
  "Updates a list of items"
  [inventory]
  (map update-item inventory))


(defn update-current-inventory
  "Helper function to get build an inventory and update the quality of it."
  []
  (let [inventory
    [(item "+5 Dexterity Vest" 10 20)
     (item "Aged Brie" 2 0)
     (item "Elixir of the Mongoose" 5 7)
     (item "Sulfuras, Hand Of Ragnaros" 0 80)
     (item "Backstage passes to a TAFKAL80ETC concert" 15 20)]]
    (update-inventory inventory)))
